In [1]:
# Import important library
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split

### Q1 Read the input file and check the data dimension

In [2]:
#Read input file and understand the data
# "default" is my dependent variable

In [3]:
df = pd.read_csv('german_credit.csv')

In [4]:
df.head()

default account_check_status  duration_in_month  \
0        0               < 0 DM                  6   
1        1    0 <= ... < 200 DM                 48   
2        0  no checking account                 12   
3        0               < 0 DM                 42   
4        1               < 0 DM                 24   

                                      credit_history  \
0  critical account/ other credits existing (not ...   
1           existing credits paid back duly till now   
2  critical account/ other credits existing (not ...   
3           existing credits paid back duly till now   
4                    delay in paying off in the past   

                        purpose  credit_amount                      savings  \
0           domestic appliances           1169  unknown/ no savings account   
1           domestic appliances           5951                 ... < 100 DM   
2  (vacation - does not exist?)           2096                 ... < 100 DM   
3              radio/television           7882                 ... < 100 DM   
4                     car (new)           4870                 ... < 100 DM   

    present_emp_since  installment_as_income_perc  \
0       .. >= 7 years                           4   
1  1 <= ... < 4 years                           2   
2  4 <= ... < 7 years                           2   
3  4 <= ... < 7 years                           2   
4  1 <= ... < 4 years                           3   

                   personal_status_sex  ... present_res_since  \
0                        male : single  ...                 4   
1  female : divorced/separated/married  ...                 2   
2                        male : single  ...                 3   
3                        male : single  ...                 4   
4                        male : single  ...                 4   

                                            property age  \
0                                        real estate  67   
1                                        real estate  22   
2                                        real estate  49   
3  if not A121 : building society savings agreeme...  45   
4                              unknown / no property  53   

   other_installment_plans   housing credits_this_bank  \
0                     none       own                 2   
1                     none       own                 1   
2                     none       own                 1   
3                     none  for free                 1   
4                     none  for free                 2   

                           job people_under_maintenance  \
0  skilled employee / official                        1   
1  skilled employee / official                        1   
2         unskilled - resident                        2   
3  skilled employee / official                        2   
4  skilled employee / official                        2   

                                   telephone foreign_worker  
0  yes, registered under the customers name             yes  
1                                       none            yes  
2                                       none            yes  
3                                       none            yes  
4                                       none            yes  

[5 rows x 21 columns]

In [5]:
df.shape

(1000, 21)

In [6]:
df.columns

Index(['default', 'account_check_status', 'duration_in_month',
       'credit_history', 'purpose', 'credit_amount', 'savings',
       'present_emp_since', 'installment_as_income_perc',
       'personal_status_sex', 'other_debtors', 'present_res_since', 'property',
       'age', 'other_installment_plans', 'housing', 'credits_this_bank', 'job',
       'people_under_maintenance', 'telephone', 'foreign_worker'],
      dtype='object')

In [7]:
df.isnull().sum()

default                       0
account_check_status          0
duration_in_month             0
credit_history                0
purpose                       0
credit_amount                 0
savings                       0
present_emp_since             0
installment_as_income_perc    0
personal_status_sex           0
other_debtors                 0
present_res_since             0
property                      0
age                           0
other_installment_plans       0
housing                       0
credits_this_bank             0
job                           0
people_under_maintenance      0
telephone                     0
foreign_worker                0
dtype: int64

### Q2.Prepare the model data by converting non-numeric to dummy ( 1 Marks)
##### Hint: Use get_dummies

In [8]:
new_df = pd.get_dummies(df, drop_first = True)

In [9]:
new_df.shape

(1000, 49)

In [10]:
new_df.head(5)

default  duration_in_month  credit_amount  installment_as_income_perc  \
0        0                  6           1169                           4   
1        1                 48           5951                           2   
2        0                 12           2096                           2   
3        0                 42           7882                           2   
4        1                 24           4870                           3   

   present_res_since  age  credits_this_bank  people_under_maintenance  \
0                  4   67                  2                         1   
1                  2   22                  1                         1   
2                  3   49                  1                         2   
3                  4   45                  1                         2   
4                  4   53                  2                         2   

   account_check_status_< 0 DM  \
0                            1   
1                            0   
2                            0   
3                            1   
4                            1   

   account_check_status_>= 200 DM / salary assignments for at least 1 year  \
0                                                  0                         
1                                                  0                         
2                                                  0                         
3                                                  0                         
4                                                  0                         

   ...  property_unknown / no property  other_installment_plans_none  \
0  ...                               0                             1   
1  ...                               0                             1   
2  ...                               0                             1   
3  ...                               0                             1   
4  ...                               1                             1   

   other_installment_plans_stores  housing_own  housing_rent  \
0                               0            1             0   
1                               0            1             0   
2                               0            1             0   
3                               0            0             0   
4                               0            0             0   

   job_skilled employee / official  job_unemployed/ unskilled - non-resident  \
0                                1                                         0   
1                                1                                         0   
2                                0                                         0   
3                                1                                         0   
4                                1                                         0   

   job_unskilled - resident  \
0                         0   
1                         0   
2                         1   
3                         0   
4                         0   

   telephone_yes, registered under the customers name   foreign_worker_yes  
0                                                  1                     1  
1                                                  0                     1  
2                                                  0                     1  
3                                                  0                     1  
4                                                  0                     1  

[5 rows x 49 columns]

### Check for highly correlated variables but don't required any treatment for this use case

In [11]:
corr_matrix = new_df.corr().abs()

In [12]:
corr_matrix

default  \
default                                             1.000000   
duration_in_month                                   0.214927   
credit_amount                                       0.154739   
installment_as_income_perc                          0.072404   
present_res_since                                   0.002967   
age                                                 0.091127   
credits_this_bank                                   0.045732   
people_under_maintenance                            0.003015   
account_check_status_< 0 DM                         0.258333   
account_check_status_>= 200 DM / salary assignm...  0.044009   
account_check_status_no checking account            0.322436   
credit_history_critical account/ other credits ...  0.181713   
credit_history_delay in paying off in the past      0.012325   
credit_history_existing credits paid back duly ...  0.043722   
credit_history_no credits taken/ all credits pa...  0.144767   
purpose_business                                    0.036129   
purpose_car (new)                                   0.096900   
purpose_car (used)                                  0.099791   
purpose_domestic appliances                         0.106922   
purpose_education                                   0.020828   
purpose_furniture/equipment                         0.028058   
purpose_radio/television                            0.020971   
purpose_repairs                                     0.008016   
purpose_retraining                                  0.039281   
savings_... < 100 DM                                0.161007   
savings_100 <= ... < 500 DM                         0.022255   
savings_500 <= ... < 1000 DM                        0.070954   
savings_unknown/ no savings account                 0.129238   
present_emp_since_... < 1 year                      0.106397   
present_emp_since_1 <= ... < 4 years                0.010603   
present_emp_since_4 <= ... < 7 years                0.075980   
present_emp_since_unemployed                        0.039815   
personal_status_sex_male : divorced/separated       0.050063   
personal_status_sex_male : married/widowed          0.019630   
personal_status_sex_male : single                   0.080677   
other_debtors_guarantor                             0.055039   
other_debtors_none                                  0.000751   
property_if not A121/A122 : car or other, not i...  0.011121   
property_real estate                                0.119300   
property_unknown / no property                      0.125750   
other_installment_plans_none                        0.113285   
other_installment_plans_stores                      0.050523   
housing_own                                         0.134589   
housing_rent                                        0.092785   
job_skilled employee / official                     0.013559   
job_unemployed/ unskilled - non-resident            0.005951   
job_unskilled - resident                            0.021822   
telephone_yes, registered under the customers n...  0.036466   
foreign_worker_yes                                  0.082079   

                                                    duration_in_month  \
default                                                      0.214927   
duration_in_month                                            1.000000   
credit_amount                                                0.624984   
installment_as_income_perc                                   0.074749   
present_res_since                                            0.034067   
age                                                          0.036136   
credits_this_bank                                            0.011284   
people_under_maintenance                                     0.023834   
account_check_status_< 0 DM                                  0.022244   
account_check_status_>= 200 DM / salary assignm...           0.076455   
account_check_status_no checking account         

In [13]:
type(corr_matrix)

pandas.core.frame.DataFrame

In [14]:
filter1 = corr_matrix > 0.5

filter2 = corr_matrix != 1

In [15]:
corr_matrix[filter1 & filter2]

default  \
default                                                 NaN   
duration_in_month                                       NaN   
credit_amount                                           NaN   
installment_as_income_perc                              NaN   
present_res_since                                       NaN   
age                                                     NaN   
credits_this_bank                                       NaN   
people_under_maintenance                                NaN   
account_check_status_< 0 DM                             NaN   
account_check_status_>= 200 DM / salary assignm...      NaN   
account_check_status_no checking account                NaN   
credit_history_critical account/ other credits ...      NaN   
credit_history_delay in paying off in the past          NaN   
credit_history_existing credits paid back duly ...      NaN   
credit_history_no credits taken/ all credits pa...      NaN   
purpose_business                                        NaN   
purpose_car (new)                                       NaN   
purpose_car (used)                                      NaN   
purpose_domestic appliances                             NaN   
purpose_education                                       NaN   
purpose_furniture/equipment                             NaN   
purpose_radio/television                                NaN   
purpose_repairs                                         NaN   
purpose_retraining                                      NaN   
savings_... < 100 DM                                    NaN   
savings_100 <= ... < 500 DM                             NaN   
savings_500 <= ... < 1000 DM                            NaN   
savings_unknown/ no savings account                     NaN   
present_emp_since_... < 1 year                          NaN   
present_emp_since_1 <= ... < 4 years                    NaN   
present_emp_since_4 <= ... < 7 years                    NaN   
present_emp_since_unemployed                            NaN   
personal_status_sex_male : divorced/separated           NaN   
personal_status_sex_male : married/widowed              NaN   
personal_status_sex_male : single                       NaN   
other_debtors_guarantor                                 NaN   
other_debtors_none                                      NaN   
property_if not A121/A122 : car or other, not i...      NaN   
property_real estate                                    NaN   
property_unknown / no property                          NaN   
other_installment_plans_none                            NaN   
other_installment_plans_stores                          NaN   
housing_own                                             NaN   
housing_rent                                            NaN   
job_skilled employee / official                         NaN   
job_unemployed/ unskilled - non-resident                NaN   
job_unskilled - resident                                NaN   
telephone_yes, registered under the customers n...      NaN   
foreign_worker_yes                                      NaN   

                                                    duration_in_month  \
default                                                           NaN   
duration_in_month                                                 NaN   
credit_amount                                                0.624984   
installment_as_income_perc                                        NaN   
present_res_since                                                 NaN   
age                                                               NaN   
credits_this_bank                                                 NaN   
people_under_maintenance                                          NaN   
account_check_status_< 0 DM                                       NaN   
account_check_status_>= 200 DM / salary assignm...                NaN   
account_check_status_no checking account                          NaN   
credit_history_critical a

# Observations

1. duration_in_month and credit_amount are moderately highly correlated (0.624984)

2. credit_history_critical account/ other credits existing (not at this bank) and credits_this_bank are moderately highly correlated (0.501364) 

3. credit_history_existing credits paid back duly till now and credits_this_bank are moderately negatively correlated (0.540354) 

4. housing_rent and housing_own are moderately highly negatively correlated (0.735968)

5. job_skilled employee / official and job_unskilled_resident are moderately highly negatively correlated (0.652438)

### Q3 Split Train/Test data 70:30 ratio( 1 Marks)
##### Hint:from sklearn.model_selection import train_test_split

In [16]:
X = new_df.drop('default', axis = 1)

y = new_df['default']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.30, random_state = 7)

### Q4 Build Random Forest Model( 2 Marks)
#### Hint:from sklearn.ensemble import RandomForestClassifier using n_jobs=2,n_estimators=500,criterion="entropy",random_state=9999

In [17]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report, precision_score, recall_score

In [18]:
rfc = RandomForestClassifier(n_jobs = 2, n_estimators = 500, criterion = 'entropy', random_state = 9999)

In [19]:
rfc.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='entropy',
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=500, n_jobs=2,
                       oob_score=False, random_state=9999, verbose=0,
                       warm_start=False)

### Q5 Calculate Confusion Matrix and Accuracy score (1 Marks)
##### Hint: Use confusion_matrix and accuracy_score

In [20]:
y_pred_train = rfc.predict(X_train)

In [21]:
y_pred_test = rfc.predict(X_test)

In [22]:
accuracy_score(y_train, y_pred_train)

1.0

In [23]:
accuracy_score(y_test, y_pred_test)

0.79

In [24]:
confusion_matrix(y_test, y_pred_test)

array([[202,  15],
       [ 48,  35]], dtype=int64)

In [25]:
print(classification_report(y_test, y_pred_test))

              precision    recall  f1-score   support

           0       0.81      0.93      0.87       217
           1       0.70      0.42      0.53        83

    accuracy                           0.79       300
   macro avg       0.75      0.68      0.70       300
weighted avg       0.78      0.79      0.77       300



# Observations

1. The model is performing well on the training data but not on the test data

### Q6 Show the list of the features importance( 1 Marks)

In [26]:
print (pd.DataFrame(rfc.feature_importances_, columns = ["Imp"], index = X_train.columns))

                                                         Imp
duration_in_month                                   0.086272
credit_amount                                       0.116171
installment_as_income_perc                          0.041192
present_res_since                                   0.039835
age                                                 0.091958
credits_this_bank                                   0.020832
people_under_maintenance                            0.016292
account_check_status_< 0 DM                         0.034739
account_check_status_>= 200 DM / salary assignm...  0.010112
account_check_status_no checking account            0.056920
credit_history_critical account/ other credits ...  0.020335
credit_history_delay in paying off in the past      0.010089
credit_history_existing credits paid back duly ...  0.016202
credit_history_no credits taken/ all credits pa...  0.009035
purpose_business                                    0.011797
purpose_car (new)       

### Q7 K-fold cross-validation( 2 Marks)
##### k-fold cross validation( without stratification)
##### Usually k is set as 10-20 in practical settings, depends on data set size

In [27]:
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

In [28]:
# Use below values
num_folds = 10
seed = 77

In [29]:
#Validate the Random Forest model build above using k fold

In [30]:
kf = KFold(n_splits = 10)

In [31]:
X = np.array(X)

In [32]:
fold_count = 0

for train_index, test_index in kf.split(X):
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y[train_index], y[test_index]
    
    rfc.fit(X_train, y_train)
    
    y_pred_train = rfc.predict(X_train)
    y_pred_test = rfc.predict(X_test)
    
    fold_count = fold_count + 1
    
    print('Fold Count ', fold_count)
    print('Train Accuracy', accuracy_score(y_train, y_pred_train))
    print('Test Accuracy ', accuracy_score(y_test, y_pred_test))
    print('****************************************************')

Fold Count  1
Train Accuracy 1.0
Test Accuracy  0.87
****************************************************
Fold Count  2
Train Accuracy 1.0
Test Accuracy  0.67
****************************************************
Fold Count  3
Train Accuracy 1.0
Test Accuracy  0.79
****************************************************
Fold Count  4
Train Accuracy 1.0
Test Accuracy  0.81
****************************************************
Fold Count  5
Train Accuracy 1.0
Test Accuracy  0.81
****************************************************
Fold Count  6
Train Accuracy 1.0
Test Accuracy  0.68
****************************************************
Fold Count  7
Train Accuracy 1.0
Test Accuracy  0.75
****************************************************
Fold Count  8
Train Accuracy 1.0
Test Accuracy  0.78
****************************************************
Fold Count  9
Train Accuracy 1.0
Test Accuracy  0.75
****************************************************
Fold Count  10
Train Accuracy 1.0
Test Accurac

In [33]:
#Calculate Mean score

In [34]:
scores = cross_val_score(rfc, X, y, cv = 10, scoring = 'accuracy')

In [35]:
scores

array([0.81, 0.69, 0.72, 0.78, 0.78, 0.73, 0.75, 0.83, 0.73, 0.77])

In [36]:
scores.mean()

0.759

In [37]:
# Calculate score standard deviation using std()

In [38]:
scores.std()

0.04085339643163101

# Q8 Build Ada boosting model, check the accuracy, and print confusion matrix ( 1 Marks)

In [39]:
from sklearn.ensemble import AdaBoostClassifier, GradientBoostingClassifier

In [40]:
X = new_df.drop('default', axis = 1)

y = new_df['default']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.30, random_state = 7)

In [42]:
model = AdaBoostClassifier()

In [43]:
model.fit(X_train, y_train)

AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None, learning_rate=1.0,
                   n_estimators=50, random_state=None)

In [44]:
y_pred_train = model.predict(X_train)

In [45]:
y_pred_test = model.predict(X_test)

In [46]:
accuracy_score(y_train, y_pred_train)

0.7785714285714286

In [47]:
accuracy_score(y_test, y_pred_test)

0.7466666666666667

In [48]:
print(confusion_matrix(y_test, y_pred_test))

[[184  33]
 [ 43  40]]


# Q9. Build Gradient boosting model, check the accuracy, and print confusion matrix ( 1 Marks)
percentage of correct predictions and Calculate sensitivity (or True Positive Rate or Recall) and Precision.
( 1 Marks)

In [49]:
model = GradientBoostingClassifier()

In [50]:
model.fit(X_train, y_train)

GradientBoostingClassifier(criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='deviance', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=100,
                           n_iter_no_change=None, presort='auto',
                           random_state=None, subsample=1.0, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)

In [51]:
y_pred_train = model.predict(X_train)

In [52]:
y_pred_test = model.predict(X_test)

In [53]:
accuracy_score(y_train, y_pred_train)

0.8942857142857142

In [54]:
accuracy_score(y_test, y_pred_test)

0.7966666666666666

In [55]:
precision_score(y_test, y_pred_test)

0.671875

In [56]:
recall_score(y_test, y_pred_test)

0.5180722891566265

In [57]:
print(confusion_matrix(y_test, y_pred_test))

[[196  21]
 [ 40  43]]


### Bootstrapping ( Bonus)
##### Given a dataset of size n, a bootstrap sample is created by sampling n instances uniformly from the data (with/without replacement)
##### Create a model with each bootstrap sample and validate it with the test set
##### Final result is calculated by averaging the accuracy of models

In [146]:
# Number of iterations for bootstrapping
bootstrap_iteration = 10
accuracy = []

In [ ]:
from sklearn.utils import resample
from sklearn.metrics import accuracy_score

for i in range(bootstrap_iteration):
    X_, y_ = resample(X_train, y_train)
    rfm.fit(X_, y_)
    y_pred = rfm.predict(X_test)
    
    acc = accuracy_score(y_pred, y_test)
    accuracy.append(acc)

In [ ]:
accuracy = np.array(accuracy)
print('Accuracy Score')
print('Avearge: ', accuracy.mean())
print('Standard deviation: ', accuracy.std())